Replace dense matrix expansion with a tensordot implementation

josh146 opened the pull request:
Currently, to implement multi-qubit expectation values on `forest.wavefunction` and `forest.numpy_wavefunction`, the matrix representing the n-qubit observable is expanded from size 2^n x 2^n to size 2^N x 2^N to act on the full N qubit subsytem.

To make this computation more efficient, the `expand()` method has been replaced with a `mat_vec_product()` method, which uses `np.tensordot` to more efficiently perform the matrix-vector multiplication, without requiring an expansion of the dense matrix to match the vector size.

This is inspired by https://github.com/XanaduAI/pennylane/pull/239.

msohaibalam commented on pennylane_forest/numpy_wavefunction.py:
```diff
@@ -100,50 +100,31 @@ def ev(self, A, wires):
+        mat = np.reshape(mat, [2] * len(wires) * 2)
+        vec = np.reshape(vec, [2] * self.num_wires)
+        axes = (np.arange(len(wires), 2 * len(wires)), wires)
+        tdot = np.tensordot(mat, vec, axes=axes)
```
I can't grok these 4 lines, my brain is feeble. Maybe add a comment or two saying what these steps are for exactly? Or rather, why this particular reshaping and identification of axes creates the desired tensor contraction.

josh146 replied:
Added comments :)

msohaibalam replied:
Awesome!

josh146 commented on pennylane_forest/wavefunction.py:
```diff
@@ -164,50 +164,31 @@ def ev(self, A, wires):
+        perm = wires + unused_idxs
+        inv_perm = np.argsort(perm) # argsort gives inverse permutation
+        state_multi_index = np.transpose(tdot, inv_perm)
+        return np.reshape(state_multi_index, 2 ** self.num_wires)
```
Good idea :+1: I did it this way as it is not used by the other devices (QVM, QPU), but I suppose it doesn't matter - they can just ignore it.

josh146 commented on pennylane_forest/numpy_wavefunction.py:
```diff
@@ -100,50 +100,31 @@ def ev(self, A, wires):
+        perm = wires + unused_idxs
+        inv_perm = np.argsort(perm) # argsort gives inverse permutation
+        state_multi_index = np.transpose(tdot, inv_perm)
+        return np.reshape(state_multi_index, 2 ** self.num_wires)
```
Hey @msohaibalam, it works for me - are you also modifying A to be an operator that acts on only 2 qubits? i.e.,

```python
# cook up some matrix and vector
A = np.kron(z, h) # 2 qubit operator Z x H
v = np.kron(np.kron(ket1, ket0), ket0) # 3 qubit state

# Apply A = Z x H to qubits 0 and 1
# i.e., A = (Z x H x I).[0, 0, 0, 0, 1, 0, 0, 0]
res = dev_numpy.mat_vec_product(A, v, wires=[0, 1])

# do the densce matrix multiplication manually
res2 = np.kron(A, np.identity(2)) @ v

print(np.allclose(res, res2))
```

msohaibalam replied:
Ah, so I was actually wondering about the case where A is an operator acting on 3 qubits. I suppose that's not the intended functionality here (and that makes sense). I would suggest making this a bit more explicit. Maybe just add a comment here saying that the matrix to apply should have the corresponding dimensionality of the sub-system, and adding an error catching mechanism like
```
num_wires = len(wires)
assert mat.shape == (2**num_wires, 2**num_wires)
```

josh146 replied:
Good point, added the input validation :slightly_smiling_face:

josh146 commented on tests/test_numpy_wavefunction.py:
```diff
@@ -20,106 +20,6 @@
-            @ np.kron(SWAP, np.kron(I, I))
-        )
-        self.assertAllAlmostEqual(res, expected, delta=tol)
-
```
:+1: I'll add https://github.com/rigetti/pennylane-forest/pull/17#discussion_r308577580 as a test

msohaibalam replied:
How about choosing the name `test_device.py` instead of `test_mat_vec_product.py`? That way, any future common methods could also be tested in the same file, and it also satisfies the similar pattern of `test_xx` where `xx` is some sort of simulator/device (with the exception of `test_gradients.py` perhaps)

josh146 replied:
Good call, I've just made the change (`test_device.py`) as you suggested

josh146 commented:
Hi @msohaibalam, I've addressed your comments above! Semaphore for this PR seems to be suffering from the same problem as PR #18; the tests are failing with a Quilc TimeoutError?

msohaibalam commented:
@josh146 Added a few more comments. The QuilC error has been resolved, so I pushed another build for the tests, which seem to be passing now.
